MutatingAdmissionPolicy implementation

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implmentation of mutating admission policy with major missing features:

1. [x] API types (@cici37)
2. [x] Generation of patch from CEL (@jiahuif)
4. [x] SSA patch application
5. [x] Reinvocation
6. [x] Unit Tests
7. [ ] Integration Tests
8. [x] Feature gate
9. [x] Plugin Wiring

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

/cc @cici37 
/cc @jiahuif 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

alexzielenski committed:
refactor: expose SchemaRelativeURL for OpenAPI

need this to know when to invalidate the cache

cici37 committed:
Manual changes

alexzielenski committed:
refactor: provide a client to dispatcher constructor

alexzielenski committed:
refactor: add Run to dispatcher

to be used by mutating dispatcher

jiahuif committed:
MAP: compile policy (#5)

* WIP compilation.

* compile during creation not invocation.

alexzielenski committed:
refactor: refactor compilation MAP for real API types

and account for the fact we will need separate evaluator for each mutation to handle reinvocation

MariamFahmy98 commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -1169,3 +1169,234 @@ type MatchCondition struct {
+	// ParamKind specifies the kind of resources used to parameterize this policy.
+	// If absent, there are no parameters for this policy and the param CEL variable will not be provided to validation expressions.
+	// If ParamKind refers to a non-existent kind, this policy definition is mis-configured and the FailurePolicy is applied.
+	// If paramKind is specified but paramRef is unset in ValidatingAdmissionPolicyBinding, the params variable will be null.
```
```suggestion
	// If paramKind is specified but paramRef is unset in MutatingAdmissionPolicyBinding, the params variable will be null.
```

cici37 replied:
Thanks for the review. Updated :)

MariamFahmy98 commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -1169,3 +1169,234 @@ type MatchCondition struct {
+	MatchConstraints *MatchResources
+
+	// Mutations contain CEL expressions which is used to apply the mutation.
+	// Mutations may be empty; a minimum of one Mutations is required.
```
may not be?

cici37 replied:
Updated.

deads2k commented on staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go:
```diff
@@ -663,3 +663,249 @@ const (
+	// * mutations that use this option may be reordered to minimize the number of additional invocations.
+	// * to validate an object after all mutations are guaranteed complete, use a validating admission policy instead.
+	//
+	// Defaults to "Never".
```
I prefer to default to IfNeeded so that by default every check will be able to see mutations made by other admission plugins.

alexzielenski committed:
refactor: straighten error handling in policy dispatcher

and resolve linter issues

This makes it more clear how we handle errors for policy plugins. Two types of errors: policy errors, and internal error. Internal error raises internal k8s status immediately and aborts operation. Policy error is handled according to the failure policy of the policy raising the error.

alexzielenski committed:
feature: register mutatingadmissionpolicy plugin

alexzielenski committed:
feature: add mutatingadmissionpolicy plugin pkg

alexzielenski committed:
bugfix: fix feature gate not being registered in genericapiserver

jiahuif committed:
generated: ./hack/update-codegen.sh

jiahuif commented:
re-based due to introduction of workspace.

leilajal commented:
/triage accepted

liangyuanpeng commented on pkg/apis/admissionregistration/types.go:
```diff
@@ -1163,3 +1163,234 @@ type MatchCondition struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.30
```
today,it's 1.31

dims commented:
@alexzielenski  please rebase!

cici37 commented:
/close 
in favor of https://github.com/kubernetes/kubernetes/pull/126497

Thanks
